pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -0,0 +1,864 @@
+			// L7 policies are a deny
+			return nil
+		}
+		match := &workloadapi.RBACPolicyRuleMatch{
+			Namespaces:       whenMatch("source.namespace", when, false, stringToMatch),
+			Principals:       whenMatch("source.principal", when, false, stringToMatch),
+			SourceIps:        whenMatch("source.ip", when, false, stringToIP),
+			DestinationPorts: whenMatch("destination.port", when, false, stringToPort),
+			DestinationIps:   whenMatch("destination.ip", when, false, stringToIP),
+		}
+		rules = append(rules, &workloadapi.RBACPolicyRules{Matches: []*workloadapi.RBACPolicyRuleMatch{match}})
+	}
+	for _, when := range rule.When {
+		match := &workloadapi.RBACPolicyRuleMatch{
+			NotNamespaces:       whenMatch("source.namespace", when, true, stringToMatch),
+			NotPrincipals:       whenMatch("source.principal", when, true, stringToMatch),
+			NotSourceIps:        whenMatch("source.ip", when, true, stringToIP),
+			NotDestinationPorts: whenMatch("destination.port", when, true, stringToPort),
+			NotDestinationIps:   whenMatch("destination.ip", when, true, stringToIP),
+		}
+		rules = append(rules, &workloadapi.RBACPolicyRules{Matches: []*workloadapi.RBACPolicyRuleMatch{match}})
+	}
+	return rules
+}
+
+var l4WhenAttributes = sets.New(
+	"source.ip",
+	"source.namespace",
+	"source.principal",
+	"destination.ip",
+	"destination.port",
+)
+
+func whenMatch[T any](s string, when *v1beta1.Condition, invert bool, f func(v []string) []T) []T {
+	if when.Key != s {
+		return nil
+	}
+	if invert {
+		return f(when.Values)
+	}
+	return f(when.NotValues)
+}
+
+func stringToMatch(rules []string) []*workloadapi.StringMatch {
+	res := make([]*workloadapi.StringMatch, 0, len(rules))
+	for _, v := range rules {
+		var sm *workloadapi.StringMatch
+		switch {
+		case v == "*":
+			sm = &workloadapi.StringMatch{MatchType: &workloadapi.StringMatch_Presence{}}
+		case strings.HasPrefix(v, "*"):
+			sm = &workloadapi.StringMatch{MatchType: &workloadapi.StringMatch_Suffix{
+				Suffix: strings.TrimPrefix(v, "*"),
+			}}
+		case strings.HasSuffix(v, "*"):
+			sm = &workloadapi.StringMatch{MatchType: &workloadapi.StringMatch_Prefix{
+				Prefix: strings.TrimSuffix(v, "*"),
+			}}
+		default:
+			sm = &workloadapi.StringMatch{MatchType: &workloadapi.StringMatch_Exact{
+				Exact: v,
+			}}
+		}
+		res = append(res, sm)
+	}
+	return res
+}
+
+func stringToPort(rules []string) []uint32 {
+	res := make([]uint32, 0, len(rules))
+	for _, m := range rules {
+		p, err := strconv.ParseUint(m, 10, 32)
+		if err != nil || p > 65535 {
+			continue
+		}
+		res = append(res, uint32(p))
+	}
+	return res
+}
+
+func stringToIP(rules []string) []*workloadapi.Address {
+	res := make([]*workloadapi.Address, 0, len(rules))
+	for _, m := range rules {
+		if len(m) == 0 {
+			continue
+		}
+
+		var (
+			ipAddr        netip.Addr
+			maxCidrPrefix uint32
+		)
+
+		if strings.Contains(m, "/") {
+			ipp, err := netip.ParsePrefix(m)
```

Comment:
Not sure this case is covered in test. I think all the ones with cidrs are part of L7 policy

Reply:
I improve the tests